### PR TITLE
Formatting changes to enable more Ruff Rules

### LINKED ISSUE #213: Add more checks to Ruff for cleaner code
Currently Ruff is only checking Pyflakes and pycodestyle rules.  Other projects and mozilla check many more rules that could help keep our code cleaner.  See: https://github.com/mozilla/mozanalysis/pull/211#discussion_r1506714575

### PR DESCRIPTION
**Pull request by jaredsnyder**
fixes #213

### CODE REVIEW

**Comment by jaredsnyder** [comment]
Ruff does not yet support Sphinx-style docstrings :upside_down_face: https://github.com/astral-sh/ruff/issues/6606

**Comment by jaredsnyder** [comment]
Removed the docstring check because sphinx isn't yet supported.  Left the updates though because when sphinx is supported it'll be less work to modify them than to do it over again

File: ruff.toml
Comment (mikewilli) [inline ruff.toml:5]: I'd suggest adding `RUF005` to the ignore list, as this could lead to potential issues with numpy per [this thread](https://github.com/astral-sh/ruff/issues/2142).

This affects the changes where we change from concatenating lists to using the unpacking operator `*`, so if you adopt this suggestion we should also revert those changes.

File: src/mozanalysis/config.py
Comment (mikewilli) [inline src/mozanalysis/config.py:6]: I thought it was odd to see `List` here because the `"UP"` lint rules should be suggesting "Use `list` instead of `List` for type annotation", and in fact I see that when I run `ruff` locally on main. I went to look at why [CI is passing](https://app.circleci.com/pipelines/github/mozilla/mozanalysis/845/workflows/807f45a9-919d-44f2-b4d6-366b119a9321/jobs/2572) with this still in here and it kind of looks like ruff isn't running? Maybe I'm just missing something..

Even more strange, I see that other files (like experiment.py) have removed `List`, `Dict`, etc. in favor of `list`, `dict`, etc. as expected.

None of this is a big problem, but I'm certainly confused and a little concerned about what's going on.

File: ruff.toml
Comment (jaredsnyder) [reply ruff.toml:5]: Ah good catch, will do

File: src/mozanalysis/config.py
Comment (jaredsnyder) [reply src/mozanalysis/config.py:6]: Weirdly when I run locally ruff does not flag this and it runs fine, so maybe there is some sort of weird environment thing happening?

Comment (mikewilli) [reply src/mozanalysis/config.py:6]: Just to clarify the behavior I'm seeing... The `UP` rule error I quoted above appears for some files but not others. I don't remember if it was different last week when I looked (I suspect not), but right now I don't see the lint error for `config.py` where I commented, but if I go to `main` where the formatting changes haven't been made yet and run `ruff` with `UP` rules, it flags other files (the ones that you've already fixed). I don't see any `# noqa` or other indications that the rule should be ignored in `config.py`, so I don't know why this is happening.

That said, it's probably not worth blocking on this odd behavior. I'll finish reviewing the rest of the PR now that I'm back.

File: src/mozanalysis/metrics/__init__.py
Comment (mikewilli) [inline src/mozanalysis/metrics/__init__.py:77]: This change seems unnecessary to me (and possibly less obvious what it's doing), but it is a little more succinct if you're familiar with `!r`/`!s`/`!a`. What do you think? It's RUF010 just for reference.

File: tests/bayesian_stats/test_survival_func.py
Comment (mikewilli) [inline tests/bayesian_stats/test_survival_func.py:91]: Neat!

**Review by mikewilli — APPROVED** [review]
Looks good to me! Mostly converting `.format()` strings to f-strings, and sorting imports, but still nice to have.

Might be nice to get some DS eyes on this since this is mostly a DS codebase, but I'll leave that up to you. Thanks!

File: src/mozanalysis/metrics/__init__.py
Comment (jaredsnyder) [reply src/mozanalysis/metrics/__init__.py:77]: I like it but I'm not married to it.  the `!r` format is snappy but it is a bit hard to figure out what is going on if you haven't seen it before

File: tests/bayesian_stats/test_survival_func.py
Comment (jaredsnyder) [reply tests/bayesian_stats/test_survival_func.py:91]: 😁

File: src/mozanalysis/config.py
Comment (jaredsnyder) [reply src/mozanalysis/config.py]: Lol and now the CI is flagging the format thing, fixed

File: src/mozanalysis/metrics/__init__.py
Comment (mikewilli) [reply src/mozanalysis/metrics/__init__.py:77]: I guess if ruff is doing it automatically for us it doesn't hurt. I think I'd lean slightly in favor of keeping it.

File: src/mozanalysis/config.py
Comment (mikewilli) [reply src/mozanalysis/config.py]: Oh nice work! Still seems strange to me that it fixed some but not others, but glad you got it figured out.

Comment (jaredsnyder) [reply src/mozanalysis/config.py]: Yeah it is super weird, had to dig through issues on Ruff's github to finally figure it out 🙃

File: ruff.toml
Comment (danielkberry) [inline ruff.toml:6]: Why are we ignoring `E741`?

File: setup.py
Comment (danielkberry) [inline setup.py:14]: Why are we pinning the ruff version?

File: src/mozanalysis/config.py
Comment (danielkberry) [inline src/mozanalysis/config.py:19]: Tbh I find this and the other union type changes introduced in PEP 604 harder to read but I probably just need to get with the times. 

However, it looks like this rule, [UP007](https://docs.astral.sh/ruff/rules/non-pep604-annotation/), is marked unsafe for projects using runtime type annotations like Pydantic. Pydantic is in mozanalysis's dependency chain (metric-config-parser > mozilla-nimbus-schemas > pydantic). Is this rule safe for us to use?

File: src/mozanalysis/frequentist_stats/bootstrap.py
Comment (danielkberry) [inline src/mozanalysis/frequentist_stats/bootstrap.py:240]: Wow, I'm impressed at the scope of the change made by this tool. Effectively, this is changing the type of that parameter from `List[float]` to `List[float] | None`, is that ok?

File: ruff.toml
Comment (jaredsnyder) [reply ruff.toml:6]: It was ignored back when we were using black so I kept it when we transitioned to RUFF.  I can try turning it off if you think it's unnecessary

File: setup.py
Comment (jaredsnyder) [reply setup.py:14]: My thinking was that it would be very annoying if the linting got out of sync because the versions were different, or if the CI was using a different version than local.  Don't know if that is a reasonable fear though

File: src/mozanalysis/config.py
Comment (jaredsnyder) [reply src/mozanalysis/config.py:19]: If I'm reading the documentation right it sounds likes it's only a concern before Python 3.10, which is our minimum version for the library.

File: src/mozanalysis/frequentist_stats/bootstrap.py
Comment (jaredsnyder) [reply src/mozanalysis/frequentist_stats/bootstrap.py:240]: I think the rule that's triggering is that you shouldn't have mutable types as defaults, so we could always change the default to a tuple if you don't like this fix

File: setup.py
Comment (danielkberry) [reply setup.py:14]: That makes sense to me. Then we can manually update the version independently of other changes we're making. Something similar happened on my most recent Jetstream PR (where linting changes ended up being bundled with functional changes and it was difficult to review).

File: src/mozanalysis/frequentist_stats/bootstrap.py
Comment (danielkberry) [reply src/mozanalysis/frequentist_stats/bootstrap.py:240]: Definitely agree that mutable types as a default is a footgun. I suppose so long as this change passes all our tests, I'm good with it.

File: setup.py
Comment (mikewilli) [reply setup.py:14]: I worry that this will lead to mozanalysis being stuck on 0.3.1 forever. I think we should remove this pinning as part of the work done on dependency management that comes out of [this ADR](https://github.com/mozilla/mozanalysis/pull/216). That will allow it update like any other package, and we'll just have to be intentional about how we update packages to avoid scenarios like the one @danielkberry described.

Comment (jaredsnyder) [reply setup.py:14]: This is a really good point, I added a comment to #202 so we don't forget

**Pull request merged**
